lib/mixlib/install/script_generator.rb

```diff
@@ -0,0 +1,217 @@
+      # @return [String] shell variable lines
+      # @api private
+      def install_command_vars_for_powershell
+        [
+          shell_var("chef_omnibus_root", root),
+          shell_var("msi", "$env:TEMP\\chef-#{version}.msi")
+        ].tap { |vars|
+          if install_msi_url
+            vars << shell_var("chef_msi_url", install_msi_url)
+          else
+            vars << shell_var("chef_metadata_url", windows_metadata_url)
+            vars << shell_var("pretty_version", Util.pretty_version(version))
+            vars << shell_var("version", version)
+          end
+        }.join("\n")
+      end
+
+      def validate_opts!(opt)
+        err_msg = ["#{opt} is not a valid option",
+                   "valid options are #{VALID_INSTALL_OPTS.join(" ")}"].join(",")
+        fail ArgumentError, err_msg unless VALID_INSTALL_OPTS.include?(opt.to_s)
+      end
+
+      def parse_opts(opts)
+        opts.each do |opt, setting|
+          validate_opts!(opt)
+          case opt.to_s
+          when "project", "endpoint"
+            self.endpoint = metadata_endpoint_from_project(setting)
+          else
+            send("#{opt.to_sym}=", setting)
+          end
+        end
+      end
+
+      def shell_code_from_file(vars)
+        fn = File.join(
+          File.dirname(__FILE__),
+          %w[.. .. .. support],
+          "install_command"
+        )
+        Util.shell_code_from_file(vars, fn, powershell,
+                                  http_proxy: http_proxy, https_proxy: https_proxy)
+      end
+
+      # Builds a shell variable assignment string for the required shell type.
+      #
+      # @param name [String] variable name
+      # @param value [String] variable value
+      # @return [String] shell variable assignment
+      # @api private
+      def shell_var(name, value)
+        Util.shell_var(name, value, powershell)
+      end
+
+      # @return the correct Chef Omnitruck API metadata endpoint, based on project
+      def metadata_endpoint_from_project(project = nil)
+        if project.nil? || project.downcase == "chef"
+          "metadata"
+        else
+          "metadata-#{project.downcase}"
+        end
+      end
+
+      def windows_metadata_url
+        base = if omnibus_url =~ %r{/install.sh$}
+          "#{File.dirname(omnibus_url)}/"
+        end
+
+        url = "#{base}#{endpoint}"
+        url << "?p=windows&m=x86_64&pv=2008r2" # same package for all versions
+        url << "&v=#{CGI.escape(version.to_s.downcase)}"
+        url << "&prerelease=true" if prerelease
+        url << "&nightlies=true" if nightlies
+        url
+      end
+
+      # Conditionally prefixes a command with a sudo command.
+      #
+      # @param command [String] command to be prefixed
+      # @return [String] the command, conditionaly prefixed with sudo
+      # @api private
+      def sudo(script)
+        use_sudo ? "#{sudo_command} #{script}" : script
+      end
+    end
+  end
+end
```

Comment:
Totally agreed. This will come in a future PR when we are looking into ScriptGenerator. For now we just moved the functionality as is until we have all the underlying things ready.